### PR TITLE
TEAL spec

### PR DESCRIPTION
**Pull request by algobolson**
"Transaction Execution Approval Language" (TEAL) spec

### CODE REVIEW

File: dev/TEAL_opcodes.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+- pop a value from the stack and store to scratch space
+
+## bnz
+
+- Opcode: 0x40 {0..0x7fff forward branch offset, big endian}
+- Pops: *... stack*, uint64
+- Pushes: _None_
+- branch if value is not zero
+
+for a bnz instruction at `pc`, if the last element of the stack is not zero then branch to instruction at `pc + 3 + N`, else proceed to next instruction at `pc + 3`
```
Comment (algoradam) [inline dev/TEAL_opcodes.md]: We reject if any jump would land not at the start of an instruction, right? That should probably be mentioned here.

Comment (derbear) [inline dev/TEAL_opcodes.md]: May want to specify that jumps must be instruction-aligned here. (If unaligned, program returns false.)

File: dev/TEAL_opcodes.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+## rand
+
+- Opcode: 0x05 
+- Pops: _None_
+- Pushes: uint64
+- push random uint64 to stack
+- **Cost**: 3
+
+Random number generator based on the ChaCha20 algorithm. Seeded with the previous block's `Seed` value and the current transaction ID.
```
Comment (derbear) [inline dev/TEAL_opcodes.md]: Is there a rationale for seeding with the current transaction ID?

In general, we should be wary of providing this instruction--it may be theoretically impossible for the Algorand TEAL execution environment to provide applications with truly unbiased randomness.

Comment (derbear) [reply dev/TEAL_opcodes.md]: (Also might be good to add reference to ChaCha20 spec.)

File: dev/TEAL.md [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Round | uint64 |
+| 1 | MinTxnFee | uint64 |
+| 2 | MinBalance | uint64 |
+| 3 | MaxTxnLife | uint64 |
+| 4 | TimeStamp | uint64 |
```
Comment (algoradam) [inline dev/TEAL.md]: Is this the current block's timestamp or the previous block's timestamp?

File: dev/TEAL.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+int 1
+bnz safe
+err
+safe:
+pop
+```
+
+# Encoding and Versioning
+
+A program starts with a varuint declaring the version of the compiled code. Any addition, removal, or change of opcode behavior increments the version. For the most part opcode behavior should not change, addition will be infrequent (not likely more often than every three months and less often as the language matures), and removal should be very rare.
```
Comment (derbear) [inline dev/TEAL.md]: If an old opcode is removed for whatever reason, do old programs break?

File: dev/TEAL.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+date: \today
+abstract: >
+  Algorand allows transactions to be effectively signed by a small program. If the program evaluates to true then the transaction is allowed. This document defines the language and bytcode format.
+---
+
+# Transaction Execution Approval Language (TEAL)
+
+TEAL is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature.
+
+TEAL programs should be short, at most 1000 bytes including all constants and operations, and run fast as they are run in-line along with signature checking, transaction balance rule checking, and other checks during block assembly and validation.
```
Comment (derbear) [inline dev/TEAL.md]: Might want to parametrize "1000" here (e.g., `MaxProgramSize`).

File: dev/TEAL.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+
+In addition to the stack there are 256 positions of scratch space, also uint64-bytes union values, accessed by the `load` and `store` ops moving data from or to scratch space, respectively.
+
+## Execution Environment
+
+TEAL runs in Algorand nodes as part of testing a proposed transaction to see if it is valid and authorized to be committed into a block. If an authorized program executes and finishes with a single non-zero uint64 value on the stack then that program has validated the transaction it is attached to.
+
+A program is an authorized program one of two ways: The SHA512_256 hash of the program (prefixed by "Program") is equal to the transaction Sender address; OR the program (prefixed by "Program") has a valid signature or multi-signature from the transaction Sender address.
+
+The TEAL bytecode plus the length of any Args must add up to less than 1000 bytes. Each TEAL op has an associated cost estimate and the program cost estimate must total less than 20000. Most ops have an estimated cost of 1, but a few slow crypto ops are much higher.
```
Comment (derbear) [inline dev/TEAL.md]: May also want to parametrize 2000 (e.g., `MaxProgramCost`)

File: dev/ledger.md [hunk 7]
```diff
@@ -368,8 +368,9 @@ A key registration transaction additionally has the following fields:
 The cryptographic hash of the fields above is called the _transaction
 identifier_.  This is written as $\Hash(\Tx)$.
 
-A valid transaction can either be a _signed_ transaction or a _multi-signed_
-transaction.  This is determined by the _signature_ of a transaction:
+A valid transaction can either be a _signed_ transaction, a _multi-signed_
+transaction, or a _logic-signed_ transaction.
```
Comment (derbear) [inline dev/ledger.md]: It's probably worth defining logic signatures here, with this spec merge. Something like, 

```
- A logic signature is a [TEAL program](link) along with optional input arguments which returns true upon executing in the execution environment defined by the transaction, the arguments, the round, ... (insert the rest of the things that make up the execution environment). A logic signature is valid if either:
 - The logic signature is signed/multisigned by the transactions' sender, or
 - \Hash(logic signature) = \Sender.
```

Comment (algoradam) [inline dev/ledger.md]: At some point we'll want to describe the structure of the logicsig struct here (like https://github.com/algorandfoundation/specs/pull/6 does for multisig)

Comment (algoradam) [reply dev/ledger.md]: Probably worth waiting for the multisig spec PR (#6) to get merged first to avoid merge conflicts

File: dev/TEAL_opcodes.md [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+- Pops: *... stack*, []byte
+- Pushes: uint64
+- yields length of byte value
+
+## btoi
+
+- Opcode: 0x17 
+- Pops: *... stack*, []byte
+- Pushes: uint64
+- converts bytes as big endian to uint64
```
Comment (derbear) [inline dev/TEAL_opcodes.md]: Does this panic if the the length of the byte slice is higher than 8?

File: dev/TEAL.md [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+
+TEAL is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature.
+
+TEAL programs should be short, at most 1000 bytes including all constants and operations, and run fast as they are run in-line along with signature checking, transaction balance rule checking, and other checks during block assembly and validation.
+
+## The Stack
+
+The stack starts empty and contains values of either uint64 or bytes. (`bytes` implemented in Go as a []byte slice) Most operations act on the stack, popping arguments from it and pushing results to it.
+
+The maximum stack depth is currently 1000.
```
Comment (derbear) [inline dev/TEAL.md]: Is it 1000, or is it the maximum program size?

File: dev/TEAL.md (hunk 5 above)
Comment (derbear) [reply dev/TEAL.md]: To make it uniform across the document, it might make more sense just to specify LogicSigMaxSize and LogicSigMaxCost somewhere in the beginning, and then just refer to those in the document. I guess if those are ever changed, then you can update fewer places in the document.

**Comment by fabrice102** [comment]
Can you remove compiled files (html and pdf) from the PR?

File: dev/TEAL_opcodes.md (hunk 1 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: noted

File: dev/TEAL.md (hunk 3 above)
Comment (algobolson) [reply dev/TEAL.md]: current. I'm pretty sure I traced the execution and we set the new block-header TimeStamp at the beginning of running the evaluator so that it's available to scripts.

File: dev/TEAL_opcodes.md (hunk 1 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: noted

File: dev/TEAL.md (hunk 4 above)
Comment (algobolson) [reply dev/TEAL.md]: We'll need provisions for running old version programs as-if they were still in that environment. Op behavior changes need to be versioned. Transaction fields need to be versioned.

File: dev/TEAL.md (hunk 5 above)
Comment (algobolson) [reply dev/TEAL.md]: I simplified this by making the opening paragraph more vague and introductory. Later details mention numbers and consensus parameters.

File: dev/TEAL.md (hunk 6 above)
Comment (algobolson) [reply dev/TEAL.md]: noted consensus param names where appropriate

File: dev/TEAL.md (hunk 9 above)
Comment (algobolson) [reply dev/TEAL.md]: 1000, unrelated constant. It could be a consensus parameter on its own? Or for now I feel like this matters less than size and cost.

File: dev/TEAL_opcodes.md (hunk 8 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: yes, noted

File: dev/ledger.md (hunk 7 above)
Comment (algobolson) [reply dev/ledger.md]: added a section like the msig section

File: dev/TEAL.md (hunk 3 above)
Comment (algoradam) [reply dev/TEAL.md]: Gotcha. Fine either way, although we might see some weirdness with the transaction pool because the block evaluator it uses for deciding whether to add a transaction to the pool doesn't know the upcoming block's timestamp yet. (On the other hand, the pool buffers more than one block's worth of transactions, so under congestion it doesn't necessarily know the previous block's timestamp either.)

**Comment by algobolson** [comment]
I think this is ready to go. Any further changes needed?

File: dev/TEAL_opcodes.md [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+- push LogicSig.Args[3] to stack
+
+## txn
+
+- Opcode: 0x31 {uint8 transaction field index}
+- Pops: _None_
+- Pushes: any
+- push field from current transaction to stack
+
+Most fields are a simple copy of a uint64 or byte string value. `XferAsset` is the concatenation of the AssetID Creator Address (32 bytes) and the big-endian bytes of the uint64 AssetID Index for a total of 40 bytes. `SenderBalance` is the uin64 balance of the sender, with rewards factored in, at the time of execution.
```
Comment (derbear) [inline dev/TEAL_opcodes.md]: This line on `SenderBalance` is out of date.

File: dev/TEAL.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+* TEAL cannot create or change a transaction, only approve or reject.
+* TEAL cannot lookup balances of Algos or other assets. (Standard transaction accounting will apply after TEAL has run and authorized a transaction. A TEAL-approved transaction could still be invalid by other accounting rules just as a standard signed transaction could be invalid. e.g. I can't give away money I don't have.)
+* TEAL cannot access information in previous blocks. TEAL cannot access most information in other transactions in the current block. (TEAL can access fields of the transaction it is attached to and the transactions in an atomic transaction group.)
+* TEAL cannot know exactly what round the current transaction will commit in (but it is somewhere in FirstValid through LastValid).
+* TEAL cannot know exactly what time its transaction is committed. (`txn FirstValidTime` should be approximately the 'unix time' seconds since 1970-01-01 00:00:00 UTC of the block at FirstValid, but there are conditions in which this time may drift and slowly re-align to close to accurate time.)
+* TEAL cannot loop. It's branch instruction `bnz` "branch if not zero" can only branch forward so as to skip some code.
```
Comment (derbear) [inline dev/TEAL.md]: Nit: "Its"

File: dev/ledger.md [hunk 12]
```diff
@@ -411,6 +412,17 @@ transaction.  This is determined by the _signature_ of a transaction:
 
    - The multisignature version _v_ (current value is 1).
 
+- A valid logic-signed transaction's signature is the _lsig_ object containing
```
Comment (fabrice102) [inline dev/ledger.md]: Maybe recall here what makes a logic-signed tx valid

File: dev/TEAL_opcodes.md [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+## store
+
+- Opcode: 0x35 {uint8 position in scratch space to store to}
+- Pops: *... stack*, any
+- Pushes: _None_
+- pop a value from the stack and store to scratch space
+
+## bnz
+
+- Opcode: 0x40 {0..0x7fff forward branch offset, big endian}
```
Comment (fabrice102) [inline dev/TEAL_opcodes.md]: What is this notation? It is not using varuint?
Indicate that `N` is the immediate?

File: dev/ledger.md [hunk 14]
```diff
@@ -411,6 +412,17 @@ transaction.  This is determined by the _signature_ of a transaction:
...
+  the following fields:
+
+  - The logic _l_ which is versioned bytecode. (See TEAL docs for details)
+
+  - An optional single signature _sig_ of 64 bytes valid for the sender of the transaction which has signed the bytes in _l_.
```
Comment (fabrice102) [inline dev/ledger.md]: Indicate that the signature only covers the code _l_ and not the arguments

File: dev/TEAL.md [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+
+**Global Fields**
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | MinTxnFee | uint64 |
+| 1 | MinBalance | uint64 |
+| 2 | MaxTxnLife | uint64 |
+| 3 | ZeroAddress | []byte |
+| 4 | GroupSize | uint64 |
```
Comment (fabrice102) [inline dev/TEAL.md]: Should GroupSize be moved to `Transaction field` as it's specific to the transaction group (similarly to `GroupIndex`)

Alternatively maybe change to "Other fields" (rather than "Global")? In any case, need some additional text to explain what this field is.

File: dev/TEAL_opcodes.md [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | MinTxnFee | uint64 |
+| 1 | MinBalance | uint64 |
+| 2 | MaxTxnLife | uint64 |
+| 3 | ZeroAddress | []byte |
+| 4 | GroupSize | uint64 |
+
+
+## gtxn
```
Comment (fabrice102) [inline dev/TEAL_opcodes.md]: Add that for a transaction with group index i, `gtxn i fieldIndex` has the same behaviour as `txn fieldIndex`.

File: dev/TEAL.md [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+
+| Op | Description |
+| --- | --- |
+| `sha256` | SHA256 hash of value, yields [32]byte |
+| `keccak256` | Keccak256 hash of value, yields [32]byte |
+| `sha512_256` | SHA512_256 hash of value, yields [32]byte |
+| `ed25519verify` | for (data, signature, pubkey) verify the signature of ("ProgData" \|\| program_hash \|\| data) against the pubkey => {0 or 1} |
+| `+` | A plus B. Panic on overflow. |
+| `-` | A minus B. Panic if B > A. |
+| `/` | A divided by B. Panic if B == 0. |
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: uint64 division, so, truncation

File: dev/TEAL.md [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+The instruction set was designed to execute calculator-like expressions.
+What might be a one line expression with various parenthesized clauses should be efficiently representable in TEAL.
+
+Looping is not possible, by design, to ensure predictably fast execution.
+There is a branch instruction (`bnz`, branch if not zero) which allows forward branching only so that some code may be skipped.
+
+Many programs need only a few dozen instructions. The instruction set has some optimization built in. `intc`, `bytec`, and `arg` take an immediate value byte, making a 2-byte op to load a value onto the stack, but they also have single byte versions for loading values 0, 1, 2, and 3. Any program will benefit from having a few common values loaded with a smaller one byte opcode, and many programs won't even need more than those 4 values. Cryptographic hashes and `ed25519verify` are single byte opcodes with powerful libraries behind them. These operations still take more time than other ops (and this is reflected in the cost of each op and the cost limit of a program) but are efficient in compiled code space.
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: I feel like 'immediate' is a standard term when referring to opcode languages.
`bnz` takes an int16 two-byte immediate branch offset (currently limited to positive int15 values so that we could maybe extend it to signed uint16 in the future).
`gtxn` takes two immediates, the group index of the transaction it's looking at and the field to read from that transaction and push to the stack.

File: dev/TEAL.md [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+| `&` | A bitwise-and B |
+| `^` | A bitwise-xor B |
+| `~` | bitwise invert value |
+| `mulw` | A times B out to 128-bit long result as low (top) and high uint64 values on the stack |
+
+### Loading Values
+
+Opcodes for getting data onto the stack.
+
+Some of these have immediate data in the byte or bytes after the opcode.
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: I feel like it's not _quite_ regular enough for that parallel structure to work well

File: dev/TEAL.md [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+| `bytec_1` | push bytes constant 1 to stack |
+| `bytec_2` | push bytes constant 2 to stack |
+| `bytec_3` | push bytes constant 3 to stack |
+| `arg` | push LogicSig.Args[N] value to stack by index |
+| `arg_0` | push LogicSig.Args[0] to stack |
+| `arg_1` | push LogicSig.Args[1] to stack |
+| `arg_2` | push LogicSig.Args[2] to stack |
+| `arg_3` | push LogicSig.Args[3] to stack |
+| `txn` | push field from current transaction to stack |
+| `gtxn` | push field to the stack from a transaction in the current transaction group |
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: yes

File: dev/TEAL.md [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+
+# Assembler Syntax
+
+The assembler parses line by line. Ops that just use the stack appear on a line by themselves. Ops that take arguments are the op and then whitespace and then any argument or arguments.
+
+"`//`" prefixes a line comment.
+
+## Constants and Pseudo-Ops
+
+A few pseudo-ops simplify writing code. `int` and `byte` and `addr` followed by a constant record the constant to a `intcblock` or `bytecblock` at the beginning of code and insert an `intc` or `bytec` reference where the instruction appears to load that value. `addr` parses an Algorand account address base32 and converts it to a regular bytes constant.
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: only at assembly time, the address-with-checksum is parsed and a 32 byte constant is compiled

File: dev/TEAL.md [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | FirstValidTime | uint64 |
+| 4 | LastValid | uint64 |
+| 5 | Note | []byte |
+| 6 | Lease | []byte |
+| 7 | Receiver | []byte |
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: All address fields are as they are in serialized transaction records, 32 bytes without checksum. The checksum is really just for presentation in human-readable human-writable contexts.

File: dev/TEAL.md [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+byte base32 AAAA...
+byte b32 AAAA...
+byte base32(AAAA...)
+byte b32(AAAA...)
+byte 0x0123456789abcdef...
+```
+
+`int` constants may be `0x` prefixed for hex, `0` prefixed for octal, or decimal numbers.
+
+`intcblock` may be explictly assembled. It will conflict with the assembler gathering `int` pseudo-ops into a `intcblock` program prefix, but may be used in code only has explicit `intc` references. `intcblock` should be followed by space separated int constants all on one line.
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL.md]: good typo catch.
The assembler would write out an `intcblock` at the start of the program, and then the `intcblock` written into the program would run and replace the constants loaded by the first one. I don't want to support a mixed mode because I feel like it should be really rare for anyone to want to do that. All of our example scripts just use `int`/`byte`/`addr` and let the assembler manage `intcblock` and `bytecblock` and I expect that is what developers will do. The exception would be a compiled higher level language that does something like register allocation and thus arranges its own `intcblock`.

File: dev/TEAL_opcodes.md [hunk 24]
```diff
@@ -0,0 +1,447 @@
...
+- A plus B. Panic on overflow.
+
+## -
+
+- Opcode: 0x09 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A minus B. Panic if B > A.
+
+## /
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: uint64 cpu division (truncation)

File: dev/TEAL_opcodes.md [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+- A divided by B. Panic if B == 0.
+
+## *
+
+- Opcode: 0x0b 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64
+- A times B. Panic on overflow.
+
+It is worth noting that there are 10,000,000,000,000,000 micro-Algos in the total supply, or a bit less than 2^54. When doing rational math, e.g. (A * (N/D)) as ((A * N) / D) one should limit the numerator to less than 2^10 to be completely sure there won't be overflow.
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: numerator and denominator of a rational number

File: dev/TEAL_opcodes.md [hunk 26]
```diff
@@ -0,0 +1,447 @@
...
+- bitwise invert value
+
+## mulw
+
+- Opcode: 0x1d 
+- Pops: *... stack*, {uint64 A}, {uint64 B}
+- Pushes: uint64, uint64
+- A times B out to 128-bit long result as low (top) and high uint64 values on the stack
+
+## intcblock
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: the contents of the first one are forgotten after the second one runs

File: dev/TEAL_opcodes.md [hunk 27]
```diff
@@ -0,0 +1,447 @@
...
+- Pops: _None_
+- Pushes: uint64
+- push value from uint64 constants to stack by index into constants
+
+## intc_0
+
+- Opcode: 0x22 
+- Pops: _None_
+- Pushes: uint64
+- push uint64 constant 0 to stack
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: rewording to be "push constant 0 from intcblock to stack" and similar for bytec_* and bytecblock

File: dev/TEAL.md (hunk 15 above)
Comment (algobolson) [reply dev/TEAL.md]: I feel like `GroupSize` isn't a property of any one `txn` and so I moved it out into 'the environment in which the txn exists' that I called `global`.

File: dev/TEAL_opcodes.md [hunk 28]
```diff
@@ -0,0 +1,447 @@
...
+- push uint64 constant 2 to stack
+
+## intc_3
+
+- Opcode: 0x25 
+- Pops: _None_
+- Pushes: uint64
+- push uint64 constant 3 to stack
+
+## bytecblock
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: noted that subsequent calls replace the first

File: dev/TEAL_opcodes.md [hunk 29]
```diff
@@ -0,0 +1,447 @@
...
+- push LogicSig.Args[3] to stack
+
+## txn
+
+- Opcode: 0x31 {uint8 transaction field index}
+- Pops: _None_
+- Pushes: any
+- push field from current transaction to stack
+
+`txn` Fields:
```
[the remark this answers is not in the document]
Comment (algobolson) [reply dev/TEAL_opcodes.md]: added descriptions to most fields

File: dev/TEAL_opcodes.md (hunk 13 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: All of the ops that have immediate values are "Opcode: 0xAA {foo}{bar}" and ops without immediate values don't have "{foo}"

File: dev/ledger.md (hunk 14 above)
Comment (algobolson) [reply dev/ledger.md]: this already says precisely that

File: dev/ledger.md (hunk 12 above)
Comment (algobolson) [reply dev/ledger.md]: you mean that the program runs and returns true?

File: dev/ledger.md (hunk 14 above)
Comment (shaih) [reply dev/ledger.md]: After "- An optional array of byte strings _arg_ which are arguments supplied to the program in _l_." add "This value (if present) is not covered under the signature or multisignature field above."

Comment (algobolson) [reply dev/ledger.md]: noted that arg bytes are not covered by sigs

File: dev/TEAL.md [hunk 30]
```diff
@@ -14,7 +14,7 @@ TEAL programs should be short and run fast as they are run in-line along with si
 
 ## The Stack
 
-The stack starts empty and contains values of either uint64 or bytes. (`bytes` implemented in Go as a []byte slice) Most operations act on the stack, popping arguments from it and pushing results to it.
+The stack starts empty and contains values of either uint64 or bytes (`bytes` ar implemented in Go as a []byte slice). Most operations act on the stack, popping arguments from it and pushing results to it.
```
Comment (fabrice102) [inline dev/TEAL.md]: typo: "ar" -> "are"

File: dev/TEAL.md [hunk 31]
```diff
@@ -66,18 +66,18 @@ Some operations 'panic' and immediately end execution of the program.
 A transaction checked by a program that panics is not valid.
 A contract account governed by a buggy program might not have a way to get assets back out of it. Code carefully.
 
-### Arithmetic
+### Arithmetic, Logic, and Cryptographic Operations
 
 For one-argument ops, `X` is the last element on the stack, which is typically replaced by a new value.
 
 For two-argument ops, `A` is the previous element on the stack and `B` is the last element on the stack. These typically result in popping A and B from the stack and pushing the result.
```
Comment (fabrice102) [inline dev/TEAL.md]: Add a sentence for three-argument ops (for `ed25519verify`.

File: dev/TEAL.md [hunk 32]
```diff
@@ -134,45 +134,45 @@ Some of these have immediate data in the byte or bytes after the opcode.
...
+| 10 | VotePK | []byte | 32 byte address |
+| 11 | SelectionPK | []byte | 32 byte address |
+| 12 | VoteFirst | uint64 |  |
+| 13 | VoteLast | uint64 |  |
+| 14 | VoteKeyDilution | uint64 |  |
+| 15 | Type | []byte |  |
+| 16 | TypeEnum | uint64 | See table below |
+| 17 | XferAsset | uint64 | Asset ID |
+| 18 | AssetAmount | uint64 | value in Asset's units |
+| 19 | AssetSender | []byte | 32 byte address. Causes clawback of all value of asset from AssetSender if Sender is the Clawback address of the asset. |
```
Comment (fabrice102) [inline dev/TEAL.md]: Do you require that if Sender != Clawback,  then `AssetSender = 0`?

File: dev/TEAL.md [hunk 33]
```diff
@@ -134,45 +134,45 @@ Some of these have immediate data in the byte or bytes after the opcode.
...
+| 20 | AssetReceiver | []byte | 32 byte address |
+| 21 | AssetCloseTo | []byte | 32 byte address |
+| 22 | GroupIndex | uint64 | Position of this transaction within an atomic transaction group. A stand-alone transaction is implicitly element 0 in a group of 1. |
+| 23 | TxID | []byte | The computed ID for this transaction. 32 bytes. |
 
 
 Additional details in the [opcodes document](TEAL_opcodes.md#txn) on the `txn` op.
 
 **Global Fields**
```
Comment (fabrice102) [inline dev/TEAL.md]: Can you add:
"Global fields are fields that are common to all the transactions in the group. In particular it includes consensus parameters."

File: dev/TEAL_opcodes.md [hunk 34]
```diff
@@ -78,7 +78,7 @@ Ops have a 'cost' of 1 unless otherwise specified.
 - Pushes: uint64
 - A times B. Panic on overflow.
 
-It is worth noting that there are 10,000,000,000,000,000 micro-Algos in the total supply, or a bit less than 2^54. When doing rational math, e.g. (A * (N/D)) as ((A * N) / D) one should limit the numerator to less than 2^10 to be completely sure there won't be overflow.
+It is worth noting that there are 10,000,000,000,000,000 micro-Algos in the total supply, or a bit less than 2^54. When doing rational math, e.g. (A * (N/D)) as ((A * N) / D) one should limit the numerator to less than 2^10 to be completely sure there won't be overflow. Overflow is an error condition which halts execution and fails the transaction. Full precision is available from `mullw`.
```
Comment (fabrice102) [inline dev/TEAL_opcodes.md]: Add more context to what `A`, `N`, and `D` are supposed to represent.

For example:
 For example, let us assume that `A` is a transaction amount and `N` and `D` are the numerator and the denominator of some rational number. Then if (A*N) / D needs to be computed, `N` needs to be less than 2^10 to be completely sure there will be no overflow.

File: dev/TEAL.md [hunk 35]
```diff
@@ -111,21 +111,21 @@ Some of these have immediate data in the byte or bytes after the opcode.
 | --- | --- |
 | `intcblock` | load block of uint64 constants |
 | `intc` | push value from uint64 constants to stack by index into constants |
```
Comment (fabrice102) [inline dev/TEAL.md]: We've just realized that it may be confusing that the instruction is called `int` while it's actually only handling `uint64`.

File: dev/ledger.md (hunk 12 above)
Comment (fabrice102) [reply dev/ledger.md]: Yes, and the logic sig itself is either signed or the hash is equal to the sender address.

At the end of this part, you can add the explanation:

A logic-signed transaction's signature is valid if:
(1) Either the logic _l_ is signed by the sender of the transaction, or the hash of the logic _l_ is equal to the sender address.
(2) And the logic _l_ evaluates to true as defined in the [TEAL documentation](TEAL.md)

(check that the link above is well formed)

File: dev/TEAL_opcodes.md (hunk 13 above)
Comment (fabrice102) [reply dev/TEAL_opcodes.md]: What is `0..0x7fff` ?

File: dev/TEAL.md (hunk 15 above)
Comment (fabrice102) [reply dev/TEAL.md]: Ok

File: dev/TEAL.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (shaih) [reply dev/TEAL.md]: How about rewording line 61 slightly, as follows:

opcodes typically take one byte to specify, but some opcodes have immediate argument making them longer. The instruction set has some optimization built in. For example the opcodes `intc`, `bytec`, and `arg` take an immediate value byte (denoting an index), making a 2-byte op to load a value onto the stack. But there are also single byte versions for loading the values indexed by 0, 1, 2, and 3. Any program will benefit from having a few common values loaded with a smaller one byte opcode, and many programs won't even need more than those 4 values.
Cryptographic hashes and `ed25519verify` are single byte opcodes with powerful libraries behind them. These operations still take more time than other ops (and this is reflected in the cost of each op and the cost limit of a program) but are efficient in compiled code space.
Many useful programs need only a few dozen instructions.

File: dev/TEAL.md (hunk 32 above)
Comment (algobolson) [reply dev/TEAL.md]: In some sense I shouldn't describe clawback at all here. It will be properly and precisely described in other sections of the spec about how assets work. This should just be a short note to remind people of what the fields are used for.

File: dev/TEAL_opcodes.md (hunk 34 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: This whole paragraph is non-normative nice-to-have text, it doesn't describe any requirement of the system, it's just a helpful reminder. I don't want to make it any bigger. If it's distracting for not being fully precise I'd rather remove it entirely and let people educate themselves on how math on binary computers works.

File: dev/TEAL.md (hunk 35 above)
Comment (algobolson) [reply dev/TEAL.md]: and if we ever add signed ints they'll be called `sint`.
If languages can have default-immutable "variables" then we can have default-unsigned integers. I think we're clear in dozens of places that the ints are unsigned.

File: dev/TEAL_opcodes.md (hunk 13 above)
Comment (tsachiherman) [reply dev/TEAL_opcodes.md]: It's the zero to 32767 range. ( 0x7fff is the hexadecimal representation of  32767).

File: dev/ledger.md (hunk 12 above)
Comment (algobolson) [reply dev/ledger.md]: Added a couple sentences explaining what combinations of values are valid, and on the execution of a valid program.

File: dev/TEAL_opcodes.md (hunk 13 above)
Comment (algobolson) [reply dev/TEAL_opcodes.md]: I think with the expanded paragraph below this should make more sense now too.

**Comment by algobolson** [comment]
I think issues from todays meetings have been resolved. Check again?

File: dev/ledger.md [hunk 36]
```diff
@@ -543,6 +543,8 @@ This is determined by the _signature_ of a transaction:
 
   - An optional array of byte strings _arg_ which are arguments supplied to the program in _l_. (_arg_ bytes are not covered by _sig_ or _msig_)
 
+  The logic signature is valid if exactly one of _sig_ or _msig_ is a valid signature of the program by the sender of the tranaction, or if neither _sig_ nor _msig_ is set and the hash of the program is equal to the sender address. Also the program must execute and finish with a single non-zero value on the stack. See [TEAL documentation](TEAL.md) for details on program execution.
```
Comment (fabrice102) [inline dev/ledger.md]: Typo: tranaction

**Comment by shaih** [comment]
On Thu, Nov 14, 2019 at 5:41 PM algobolson <notifications@github.com> wrote:

> I think issues from todays meetings have been resolved. Check again?
>
All except the discussion for line 61 in TEAL.md, which was not addressed.

The current text is highly misleading, in that it says "but they also have
single byte versions for loading values 0, 1, 2, and 3". At the very least
is should be changed to "loading the values indexed by 0,1,2 and 3" (but
really I think that the text that I suggested should be used instead.

Thanks,

-- Shai

**Comment by derbear** [comment]
@shaih, I think I might be mistaken and reading the wrong revision, but it appears that the current text on line 61 has been updated here: https://github.com/algorandfoundation/specs/pull/12/files#diff-9a22d35d0387bd7eea4acdd996766848R61

In particular, (in my browser) this says:

> Many programs need only a few dozen instructions. The instruction set has some optimization built in. `intc`, `bytec`, and `arg` take an immediate value byte, making a 2-byte op to load a value onto the stack, but they also have single byte versions for loading the most common constant values. Any program will benefit from having a few common values loaded with a smaller one byte opcode. Cryptographic hashes and `ed25519verify` are single byte opcodes with powerful libraries behind them. These operations still take more time than other ops (and this is reflected in the cost of each op and the cost limit of a program) but are efficient in compiled code space.

Does this sound okay to you? Let us know if there's a problem with the currently-committed wording. I apologize if I'm reading from a wrong git commit. Thanks!

**Comment by shaih** [comment]
Yes that's okay, for some reasoni didn't see that text when looking at it
yesterday, it only showed me the original text.

-- Shai
